### ISSUE TITLE
🚨  HIGH Severity Vulnerability: Package unsafe for use as of v1.1.8 🚨

### ISSUE THREAD

**Issue by taylorjdawson** [indutny/node-ip#136]
**Until PR is merge to mitigate this attack vector, package should be deemed unsafe for use.**

NPM IP package vulnerable to Server-Side Request Forgery (SSRF) attacks, see [GitHub advisory](https://github.com/advisories/GHSA-78xj-cgh5-2h22) for more information.

**Comment by carnil** [comment]
CVE-2023-42282 is associated with this issue, mentioning it for cross-reference as well.

**Comment by glitch-txs** [comment]
@indutny any update on this?

**Comment by levpachmanov** [comment]
Hi @taylorjdawson @carnil @glitch-txs,

Notice that ip@2.0.0 is affected by CVE-2023-42282 - https://github.com/github/advisory-database/pull/3504

We're part of a startup called Seal Security that mitigates software vulnerabilities in older open source versions by backporting/creating standalone security patches - enabling more straightforward remediation in cases like this. We created an ip 1.1.8-sp and 2.0.0-sp1 that's vulnerability-free. As with all of our patches, it's open-source and available for free.

If relevant, check out [our GitHub repo](https://github.com/seal-community/patches/tree/main/npm/ip/) if you wish to learn more, or start using our [app](https://app.sealsecurity.io/?ref=ghnodeip).

Please feel free to reach us at [info@seal.security](mailto:info@seal.security) if you have any requests/questions.

**Comment by mukitmomin** [comment]
Looks like there is an open PR: #138 that fixes this issue. Any timeline on when it will be merged and released?

**Comment by damonholden** [comment]
Any update on this?

**Comment by aminekun90** [comment]
> Any update on this?

There is an open PR for this ... --'

**Comment by levpachmanov** [comment]
@mukitmomin @damonholden @aminekun90 notice @mnikolaus 's [comment](https://github.com/indutny/node-ip/pull/138#discussion_r1485957910) on #138 that the current PR covers only a limited number of cases and as @n0099 mentioned there are [many](https://github.com/indutny/node-ip/pull/138#issuecomment-1939758449) other options

**Comment by damonholden** [comment]
Are there any known workarounds then? We've had to move our project to critical vulnerability blockers only for this.

**Comment by aminekun90** [comment]
> Are there any known workarounds then? We've had to move our project to critical vulnerability blockers only for this.

For our project we unfortunately got rid of the library using node-ip

**Comment by levpachmanov** [comment]
@damonholden - as others suggested you might try getting rid of node-ip. Alternatively we (Seal Security) released a patch that we believe covers all the cases. You can check out or [our GitHub repo](https://github.com/seal-community/patches/tree/main/npm/ip/) and our [app](https://app.sealsecurity.io/?ref=ghnodeip) - it's free to use for open source projects.

**Comment by dchambers** [comment]
> We're part of a startup called Seal Security that mitigates software vulnerabilities in older open source versions by backporting/creating standalone security patches 

@levpachmanov, may I ask, was it also your team that reported the issue to NIST in the first place?

**Comment by levpachmanov** [comment]
@dchambers no, the credit goes to @cosmosofcyberspace AFAIK. We have only suggested updating the affected version range of the advisory (even though @G-rath did it first) and trying to help the community remediate the risk.

**Comment by Siliconvelly** [comment]
i am getting this NPM IP package vulnerable to Server-Side Request Forgery (SSRF) attacks on npm install in my react-native project from today i am unable to understand why this issue is arise but it shows me to downgrade my react native version and i am downgrading the it says upgrading the version so i researched that the **npm ip** is getting issue from nodemodule file in my project so i find and found this page so please give any solution or suggestion and fix as soon as possible.

**Comment by kellyselden** [comment]
@levpachmanov I'm curious, do you also publish the patches as forks to npm? Then it would be easy to consume as a package resolution override.

**Comment by dotboris** [comment]
I have submitted a PR (https://github.com/github/advisory-database/pull/3531) to GitHub's advisory database to change it to reflect the reality of the issue and reduce its severity.

**Comment by stali1234** [comment]
any latest update on this https://security.snyk.io/vuln/SNYK-JS-IP-6240864 ?

**Comment by SamLam140330** [comment]
It seems there is a PR (https://github.com/indutny/node-ip/pull/138)

**Comment by marcomontalbano** [comment]
Hi all! We can close this issue. The PR https://github.com/indutny/node-ip/pull/138#issuecomment-1951710634 is now merged and the `v2.0.1` has been released :tada:

**Comment by mukitmomin** [comment]
The CVE reported in the github advisory database is not written correctly. NPM does not accept version v1.1.9 as a patched version as the [existing CVE](https://github.com/advisories/GHSA-78xj-cgh5-2h22) lists affected versions are <=2.0.0. [This PR](https://github.com/github/advisory-database/pull/3553) fixes the advisory to accept v1.1.9 as a patched version as well. Any idea when/how CVE can be updated?

**Comment by ouuan** [comment]
This is not resolved in 1.1.9/2.0.1. See #143 for more details.

**Comment by abhishek-parative** [comment]
Can the maintainer/group with Write access take a look at #144?

**Comment by karthikTVS** [comment]
1.1.8-sp this patch version not found in registry [https://registry.npmjs.org/ip/-/ip-1.1.8-sp.tgz](url)

**Comment by n0099** [comment]
[Fun with IP address parsing](https://news.ycombinator.com/item?id=25545967)

**Comment by bybraveHQ** [comment]
For anyone still blocked on this: the root cause of both CVE-2023-42282 and the follow-up CVE-2024-29415 (GHSA-2p57-rm9w-gvfp, still unpatched in 2.0.1) is that `isPrivate()`/`isPublic()`/`isLoopback()` classify the **raw input string** with regexes. Any spelling the regexes didn't anticipate — octal/hex octets (`012.1.2.3`, `0x7f.1`), abbreviated IPv4 (`127.1`), integers (`2130706433`), uncompressed IPv6 (`000:0:0000::01`) — falls through and is reported as *public*. Patching individual regexes can't fix that class of bug, which is why the 2.0.1 fix was incomplete.

Since this repo looks unmaintained, I published a fork that fixes the root cause: **[@bybrave/ip2](https://github.com/bybraveHQ/ip2)** (`npm i @bybrave/ip2`).

- Every classification function parses the address into **bytes** first (same grammar as `net.isIP`); classification never touches the input string.
- Non-canonical spellings **throw** instead of being guessed at — fail closed when the answer guards outgoing requests. (`normalizeToLong()` is still there if you need to accept legacy notations explicitly.)
- Classification covers the full IANA special-purpose registries (CGNAT, TEST-NETs, benchmarking, multicast, ULA, …), and `::ffff:a.b.c.d` / NAT64 / 6to4 are classified by their embedded IPv4 address.
- Also fixes #104, #67, #105, ships TypeScript types and ESM, tested on Node 18/20/22.

Breaking changes vs 2.0.1 are listed in the README — it's v3.0.0 precisely because strict parsing is the fix.
